### PR TITLE
 ref(downloader): pass in options to ChartDownloader

### PR DESCRIPTION
**Pull request by bacongobbler**
Built on top of #5917 and #5918.

This restores the ability to pass in parameters at runtime to the ChartDownloader, enabling users to pass in parameters like the `--username` and `--password` flags. This also adds a test case to cover that scenario, as it was previously untested.

I also had to introduce a `WithMiddleware` function to the test server. That way the tests could inject HTTP basic auth headers into the server to test this PR.

cc'ing @mattfarina, as he raised this concern in https://github.com/helm/helm/pull/5918#discussion_r295825504 which this PR addresses (and adds a test case to help confirm).

### CODE REVIEW

File: pkg/downloader/chart_downloader.go
Comment (thomastaylor312) [inline pkg/downloader/chart_downloader.go]: Do we need `g` here or the `SetBasicAuth` on line 259? This isn't returning the getter anymore, so it seems like this is just for an error check

File: pkg/repo/repotest/server.go
Comment (thomastaylor312) [inline pkg/repo/repotest/server.go]: If we are enabling middleware, should we handle chaining multiple middlewares or leave it up to the user to pass a `HandlerFunc` that already has all the middleware composed?

**Review by thomastaylor312 — COMMENTED** [review]
A few small questions/discussion items. I'll come back and give a green checkmark once the dependent PR is merged

File: pkg/repo/repotest/server.go
Comment (bacongobbler) [reply pkg/repo/repotest/server.go]: I can look further into this. I started with the idea of allowing a way to append middleware to the chain but landed on this design as a first pass.

File: pkg/downloader/chart_downloader.go
Comment (bacongobbler) [reply pkg/downloader/chart_downloader.go]: Good catch. No, it's not needed any more. I'll remove those two.

File: pkg/repo/repotest/server.go
Comment (thomastaylor312) [reply pkg/repo/repotest/server.go]: Because it is a test server, I think it is ok. The question just came up as I was looking at the code

File: pkg/getter/httpgetter.go
Comment (thomastaylor312) [inline pkg/getter/httpgetter.go]: This comment should be updated to reflect that it just returns a `Getter` now

**Review by thomastaylor312 — COMMENTED** [review]
This is looking awesome! Just one exported function comment change and a rebase on `dev-v3` and it should be good to go

**Comment by bacongobbler** [comment]
done

**Review by thomastaylor312 — APPROVED** [review]

**Pull request merged**
